src/dmd/ctfe/bc_common.d

```diff
@@ -0,0 +1,1211 @@
+
+        }
+
+        return false;
+    }
+
+    this(const Imm32 imm32) pure
+    {
+        this.type.type = imm32.signed ? BCTypeEnum.i32 : BCTypeEnum.u32;
+        this.vType = BCValueType.Immediate;
+        this.imm32.imm32 = imm32.imm32;
+    }
+
+    this(const Imm64 imm64) pure
+    {
+        this.type.type = imm64.signed ? BCTypeEnum.i64 : BCTypeEnum.u64;
+        this.vType = BCValueType.Immediate;
+        this.imm64 = imm64;
+    }
+
+    this(const Imm23f imm23f) pure @trusted
+    {
+        this.type.type = BCTypeEnum.f23;
+        this.vType = BCValueType.Immediate;
+        this.imm32.imm32 = *cast(uint*)&imm23f;
+    }
+
+    this(const Imm52f imm52f) pure @trusted
+    {
+        this.type.type = BCTypeEnum.f52;
+        this.vType = BCValueType.Immediate;
+        this.imm64.imm64 = *cast(ulong*)&imm52f;
+    }
+
+    this(const BCParameter param) pure
+    {
+        this.vType = BCValueType.Parameter;
+        this.type = param.type;
+        this.paramIndex = param.idx;
+        this.stackAddr = param.pOffset;
+        this.name = param.name;
+    }
+
+    this(const StackAddr sp, const BCType type, const ushort tmpIndex = 0) pure
+    {
+        this.vType = BCValueType.StackValue;
+        this.stackAddr = sp;
+        this.type = type;
+        this.tmpIndex = tmpIndex;
+    }
+
+    this(const StackAddr sp, const BCType type, const ushort localIndex, string name) pure
+    {
+        this.vType = BCValueType.Local;
+        this.stackAddr = sp;
+        this.type = type;
+        this.localIndex = localIndex;
+        this.name = name;
+    }
+
+    this(const void* base, const short addr, const BCType type) pure
+    {
+        this.vType = BCValueType.StackValue;
+        this.stackAddr = StackAddr(addr);
+        this.type = type;
+    }
+
+    this(const HeapAddr addr, const BCType type = i32Type) pure
+    {
+        this.vType = BCValueType.HeapValue;
+        this.type = type;
+        this.heapAddr = addr;
+    }
+
+    this(const BCHeapRef heapRef) pure
+    {
+        this.vType = heapRef.vType;
+        switch (vType)
+        {
+        case BCValueType.StackValue, BCValueType.Parameter:
+            stackAddr = heapRef.stackAddr;
+            tmpIndex = heapRef.tmpIndex;
+            break;
+        case BCValueType.Local:
+            stackAddr = heapRef.stackAddr;
+            tmpIndex = heapRef.localIndex;
+            name = heapRef.name;
+            break;
+
+        case BCValueType.Temporary:
+            stackAddr = heapRef.stackAddr;
+            tmpIndex = heapRef.tmpIndex;
+            break;
+
+        case BCValueType.HeapValue:
+            heapAddr = heapRef.heapAddr;
+            break;
+
+        case BCValueType.Immediate:
+            imm32 = heapRef.imm32;
+            break;
+
+        default:
+            assert(0, "vType unsupported: " ~ enumToString(vType));
+        }
+    }
+}
+
+pragma(msg, "Sizeof BCValue: ", BCValue.sizeof);
+__gshared static immutable bcLastCond = () {
```

Comment:
these do not need `__gshared` and they feel like they should be (`static immutable`) members of their types